### PR TITLE
Downloader pattern

### PR DESCRIPTION
**Pull request by d4v1d03**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [ ] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [ ] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

**Comment by d4v1d03** [comment]
@zong-zhe, @Peefy , please point out the issues and remaining functionalities that need to be covered.

**Comment by zong-zhe** [comment]
Hi @d4v1d03 

The test case isn't passing and it looks like `GitDownloader` does not work.

image

File: pkg/downloader/downloader_test.go [hunk 1]
```diff
@@ -19,42 +19,43 @@ func getTestDir(subDir string) string {
...
+	if err := os.MkdirAll(gitTestDir, os.ModePerm); err != nil {
 		t.Fatal(err)
 	}
 
 	defer func() {
-		_ = os.RemoveAll(path_git)
+		_ = os.RemoveAll(gitTestDir)
 	}()
 
 	gitDownloader := GitDownloader{}
```
Comment (zong-zhe) [inline pkg/downloader/downloader_test.go:61]: Separate this test case for `GitDownloader` and do not write it together with `TestOciDownloader` for `OciDownloader`

Comment (d4v1d03) [reply pkg/downloader/downloader_test.go]: @zong-zhe , done the same, don't know why there's a timeout occuring for my TestGitDownloader() function, can you please have a look

**Pull request closed without merging**